File: cli/internal/connectapi/admin_diagnostics.go
package connectapi

import (
	"context"
	"strconv"
	"time"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
	"hmans.de/chatto/internal/core"
	adminv1 "hmans.de/chatto/internal/pb/chatto/admin/v1"
)

type adminDiagnosticsService struct {
	api *API
}

func (s *adminDiagnosticsService) GetSystemInfo(ctx context.Context, _ *connect.Request[adminv1.GetSystemInfoRequest]) (*connect.Response[adminv1.GetSystemInfoResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	diagnostics, err := s.api.core.GetAdminDiagnostics(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}

	return connect.NewResponse(&adminv1.GetSystemInfoResponse{
		SystemInfo:           adminSystemInfo(diagnostics),
		Projections:          adminProjectionStates(diagnostics.Projections),
		AssetCleanup:         adminAssetCleanupStatus(diagnostics.AssetCleanup),
		DurableWorkers:       adminDurableWorkerStatuses(diagnostics.DurableWorkers),
		ProjectionsAvailable: proto.Bool(diagnostics.ProjectionsAvailable),
	}), nil
}

func adminDurableWorkerStatuses(statuses []core.DurableWorkerAdminStatus) []*adminv1.AdminDurableWorkerStatus {
	out := make([]*adminv1.AdminDurableWorkerStatus, 0, len(statuses))
	for _, status := range statuses {
		out = append(out, &adminv1.AdminDurableWorkerStatus{
			Key:                   status.Key,
			Health:                adminDurableWorkerHealth(status.Health),
			PendingCount:          strconv.FormatUint(status.PendingCount, 10),
			AckPendingCount:       strconv.FormatUint(status.AckPendingCount, 10),
			WaitingCount:          strconv.FormatUint(uint64(max(status.WaitingCount, 0)), 10),
			RedeliveredCount:      strconv.FormatUint(uint64(max(status.RedeliveredCount, 0)), 10),
			LastDeliveredSequence: strconv.FormatUint(status.LastDeliveredSequence, 10),
			AckFloorSequence:      strconv.FormatUint(status.AckFloorSequence, 10),
		})
	}
	return out
}

func adminDurableWorkerHealth(health core.DurableWorkerHealth) adminv1.AdminDurableWorkerHealth {
	switch health {
	case core.DurableWorkerHealthInactive:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_INACTIVE
	case core.DurableWorkerHealthHealthy:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_HEALTHY
	case core.DurableWorkerHealthWorking:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_WORKING
	case core.DurableWorkerHealthUnconfirmed:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_UNCONFIRMED
	case core.DurableWorkerHealthStalled:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_STALLED
	case core.DurableWorkerHealthUnavailable:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_UNAVAILABLE
	default:
		return adminv1.AdminDurableWorkerHealth_ADMIN_DURABLE_WORKER_HEALTH_UNSPECIFIED
	}
}

func adminAssetCleanupStatus(status core.AssetCleanupAdminStatus) *adminv1.AdminAssetCleanupStatus {
	return &adminv1.AdminAssetCleanupStatus{
		Health:                 adminAssetCleanupHealth(status.Health),
		PendingCount:           int64(status.PendingCount),
		OldestPendingAt:        optionalTimestamp(status.OldestPendingAt),
		PassInProgress:         status.PassInProgress,
		LastPassAt:             optionalTimestamp(status.LastPassAt),
		LastSuccessfulPassAt:   optionalTimestamp(status.LastSuccessfulPassAt),
		UpdatedAt:              optionalTimestamp(status.UpdatedAt),
		LastPassFailed:         status.LastPassFailed,
		LastInspectedSequence:  strconv.FormatUint(status.LastInspectedSeq, 10),
		LatestDeletionSequence: strconv.FormatUint(status.LatestDeletionSeq, 10),
	}
}

func adminAssetCleanupHealth(health core.AssetCleanupHealth) adminv1.AdminAssetCleanupHealth {
	switch health {
	case core.AssetCleanupHealthInactive:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_INACTIVE
	case core.AssetCleanupHealthInitializing:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_INITIALIZING
	case core.AssetCleanupHealthHealthy:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_HEALTHY
	case core.AssetCleanupHealthRetrying:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_RETRYING
	case core.AssetCleanupHealthStalled:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_STALLED
	case core.AssetCleanupHealthUnavailable:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_UNAVAILABLE
	default:
		return adminv1.AdminAssetCleanupHealth_ADMIN_ASSET_CLEANUP_HEALTH_UNSPECIFIED
	}
}

func optionalTimestamp(value time.Time) *timestamppb.Timestamp {
	if value.IsZero() {
		return nil
	}
	return timestamppb.New(value)
}

func adminSystemInfo(diagnostics *core.AdminDiagnostics) *adminv1.AdminSystemInfo {
	return &adminv1.AdminSystemInfo{
		Connection: adminConnectionInfo(diagnostics.Connection),
		Account:    adminAccountInfo(diagnostics.Account),
		Nats:       adminNatsStats(diagnostics.JetStream),
		Stats:      adminServerStats(diagnostics.Stats),
	}
}

func adminProjectionStates(states []core.ProjectionAdminState) []*adminv1.AdminProjectionState {
	out := make([]*adminv1.AdminProjectionState, 0, len(states))
	for _, state := range states {
		out = append(out, adminProjectionState(state))
	}
	return out
}

func adminConnectionInfo(info *core.ConnectionInfo) *adminv1.AdminConnectionInfo {
	if info == nil {
		return &adminv1.AdminConnectionInfo{}
	}
	return &adminv1.AdminConnectionInfo{
		Connected:  info.Connected,
		ServerId:   info.ServerID,
		ServerName: info.ServerName,
		Version:    info.Version,
		MaxPayload: info.MaxPayload,
		Rtt:        info.RTT,
	}
}

func adminAccountInfo(info *core.AccountInfo) *adminv1.AdminAccountInfo {
	if info == nil {
		return nil
	}
	return &adminv1.AdminAccountInfo{
		Memory:        int64(info.Memory),
		MemoryUsed:    int64(info.MemoryUsed),
		Storage:       int64(info.Storage),
		StorageUsed:   int64(info.StorageUsed),
		Streams:       int32(info.Streams),
		StreamsUsed:   int32(info.StreamsUsed),
		Consumers:     int32(info.Consumers),
		ConsumersUsed: int32(info.ConsumersUsed),
	}
}

func adminServerStats(stats *core.ServerStats) *adminv1.AdminServerStats {
	if stats == nil {
		return nil
	}
	return &adminv1.AdminServerStats{
		UserCount:        int32(stats.UserCount),
		ChannelRoomCount: int32(stats.ChannelRoomCount),
		DmRoomCount:      int32(stats.DMRoomCount),
	}
}

func adminNatsStats(stats *core.JetStreamStats) *adminv1.AdminNatsStats {
	if stats == nil {
		return nil
	}

	streams := make([]*adminv1.AdminNatsStreamInfo, 0, len(stats.Streams))
	for _, stream := range stats.Streams {
		streams = append(streams, &adminv1.AdminNatsStreamInfo{
			Name:          stream.Name,
			Description:   stream.Description,
			Subjects:      append([]string(nil), stream.Subjects...),
			Storage:       stream.Storage,
			Messages:      int64(stream.Messages),
			Bytes:         int64(stream.Bytes),
			FirstSequence: strconv.FormatUint(stream.FirstSeq, 10),
			LastSequence:  strconv.FormatUint(stream.LastSeq, 10),
			ConsumerCount: int32(stream.ConsumerCount),
			Replicas:      int32(stream.Replicas),
			ClusterLeader: stream.ClusterLeader,
		})
	}

	consumers := make([]*adminv1.AdminNatsConsumerInfo, 0, len(stats.Consumers))
	for _, consumer := range stats.Consumers {
		consumers = append(consumers, &adminv1.AdminNatsConsumerInfo{
			Stream:                    consumer.Stream,
			Name:                      consumer.Name,
			Durable:                   consumer.Durable,
			FilterSubject:             consumer.FilterSubject,
			FilterSubjects:            append([]string(nil), consumer.FilterSubjects...),
			AckPolicy:                 consumer.AckPolicy,
			PullBased:                 consumer.PullBased,
			PushBound:                 consumer.PushBound,
			Pending:                   int64(consumer.Pending),
			AckPending:                int32(consumer.AckPending),
			Redelivered:               int32(consumer.Redelivered),
			Waiting:                   int32(consumer.Waiting),
			DeliveredConsumerSequence: strconv.FormatUint(consumer.DeliveredConsumerSeq, 10),
			DeliveredStreamSequence:   strconv.FormatUint(consumer.DeliveredStreamSeq, 10),
			AckFloorConsumerSequence:  strconv.FormatUint(consumer.AckFloorConsumerSeq, 10),
			AckFloorStreamSequence:    strconv.FormatUint(consumer.AckFloorStreamSeq, 10),
		})
	}

	return &adminv1.AdminNatsStats{
		TotalMessages:        int64(stats.TotalMessages),
		TotalBytes:           int64(stats.TotalBytes),
		TotalConsumerPending: int64(stats.TotalConsumerPending),
		TotalAckPending:      int32(stats.TotalAckPending),
		Streams:              streams,
		Consumers:            consumers,
	}
}

func adminProjectionState(state core.ProjectionAdminState) *adminv1.AdminProjectionState {
	metrics := make([]*adminv1.AdminProjectionMetric, 0, len(state.Metrics))
	for _, metric := range state.Metrics {
		metrics = append(metrics, &adminv1.AdminProjectionMetric{
			Name:  metric.Name,
			Value: metric.Value,
			Bytes: metric.Bytes,
		})
	}

	var startupDurationSeconds *float64
	if state.StartupComplete {
		startupDurationSeconds = &state.StartupDuration
	}

	return &adminv1.AdminProjectionState{
		Key:                    state.Key,
		Name:                   state.Name,
		Subjects:               append([]string(nil), state.Subjects...),
		Started:                state.Started,
		StartupDurationSeconds: startupDurationSeconds,
		LastAppliedSequence:    strconv.FormatUint(state.LastAppliedSeq, 10),
		MatchingStreamSequence: strconv.FormatUint(state.MatchingStreamSeq, 10),
		StreamLastSequence:     strconv.FormatUint(state.StreamLastSeq, 10),
		Lag:                    int64(state.Lag),
		Failed:                 state.Failed,
		FailedSequence:         strconv.FormatUint(state.FailedSeq, 10),
		Failure:                state.Failure,
		EntryCount:             state.EntryCount,
		EstimatedBytes:         state.EstimatedBytes,
		AverageEntryBytes:      state.AverageEntryBytes,
		Metrics:                metrics,
	}
}

File: apps/frontend/src/lib/api-client/adminDiagnostics.ts
import { authHeaders, createChattoClient } from './connect.js';
import { AdminDiagnosticsService } from '@chatto/api-types/admin/v1/diagnostics_connect';
import {
  AdminAssetCleanupHealth,
  AdminDurableWorkerHealth
} from '@chatto/api-types/admin/v1/diagnostics_pb';

export type AdminDiagnosticsAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type AdminSystemInfo = {
  connection: AdminConnectionInfo;
  account: AdminAccountInfo;
  accountAvailable: boolean;
  nats: AdminNatsStats;
  natsAvailable: boolean;
  stats: AdminServerStats;
  statsAvailable: boolean;
  projections: AdminProjectionState[];
  projectionsAvailable: boolean;
  assetCleanup: AdminAssetCleanupStatus;
  durableWorkers: AdminDurableWorkerStatus[];
};

export type AdminDurableWorkerStatus = {
  key: string;
  health: 'inactive' | 'healthy' | 'working' | 'unconfirmed' | 'stalled' | 'unavailable';
  pendingCount: string;
  ackPendingCount: string;
  waitingCount: string;
  redeliveredCount: string;
  lastDeliveredSequence: string;
  ackFloorSequence: string;
};

export type AdminAssetCleanupStatus = {
  available: boolean;
  health: 'unavailable' | 'inactive' | 'initializing' | 'healthy' | 'retrying' | 'stalled';
  pendingCount: number;
  oldestPendingAt: Date | null;
  passInProgress: boolean;
  lastPassAt: Date | null;
  lastSuccessfulPassAt: Date | null;
  updatedAt: Date | null;
  lastPassFailed: boolean;
  lastInspectedSequence: string;
  latestDeletionSequence: string;
};

export type AdminConnectionInfo = {
  connected: boolean;
  serverId: string;
  serverName: string;
  version: string;
  maxPayload: number;
  rtt: string;
};

export type AdminAccountInfo = {
  memory: number;
  memoryUsed: number;
  storage: number;
  storageUsed: number;
  streams: number;
  streamsUsed: number;
  consumers: number;
  consumersUsed: number;
};

export type AdminServerStats = {
  userCount: number;
  channelRoomCount: number;
  dmRoomCount: number;
};

export type AdminNatsStats = {
  totalMessages: number;
  totalBytes: number;
  totalConsumerPending: number;
  totalAckPending: number;
  streams: AdminNatsStreamInfo[];
  consumers: AdminNatsConsumerInfo[];
};

export type AdminNatsStreamInfo = {
  name: string;
  description: string;
  subjects: string[];
  storage: string;
  messages: number;
  bytes: number;
  firstSequence: string;
  lastSequence: string;
  consumerCount: number;
  replicas: number;
  clusterLeader: string;
};

export type AdminNatsConsumerInfo = {
  stream: string;
  name: string;
  durable: string;
  filterSubject: string;
  filterSubjects: string[];
  ackPolicy: string;
  pullBased: boolean;
  pushBound: boolean;
  pending: number;
  ackPending: number;
  redelivered: number;
  waiting: number;
  deliveredConsumerSequence: string;
  deliveredStreamSequence: string;
  ackFloorConsumerSequence: string;
  ackFloorStreamSequence: string;
};

export type AdminProjectionState = {
  key: string;
  name: string;
  subjects: string[];
  started: boolean;
  startupDurationSeconds: number | null;
  lastAppliedSequence: string;
  matchingStreamSequence: string;
  streamLastSequence: string;
  lag: number;
  failed: boolean;
  failedSequence: string;
  failure: string;
  entryCount: number;
  estimatedBytes: number;
  averageEntryBytes: number;
  metrics: AdminProjectionMetric[];
};

export type AdminProjectionMetric = {
  name: string;
  value: number;
  bytes: number;
};

function adminDiagnosticsClient(config: AdminDiagnosticsAPIConfig) {
  const client = createChattoClient(AdminDiagnosticsService, config);
  const headers = authHeaders(config);
  return { client, headers };
}

function assetCleanupHealth(
  health: AdminAssetCleanupHealth | undefined
): AdminAssetCleanupStatus['health'] {
  switch (health) {
    case AdminAssetCleanupHealth.INACTIVE:
      return 'inactive';
    case AdminAssetCleanupHealth.INITIALIZING:
      return 'initializing';
    case AdminAssetCleanupHealth.HEALTHY:
      return 'healthy';
    case AdminAssetCleanupHealth.RETRYING:
      return 'retrying';
    case AdminAssetCleanupHealth.STALLED:
      return 'stalled';
    case AdminAssetCleanupHealth.UNAVAILABLE:
      return 'unavailable';
    default:
      return 'unavailable';
  }
}

function durableWorkerHealth(
  health: AdminDurableWorkerHealth | undefined
): AdminDurableWorkerStatus['health'] {
  switch (health) {
    case AdminDurableWorkerHealth.INACTIVE:
      return 'inactive';
    case AdminDurableWorkerHealth.HEALTHY:
      return 'healthy';
    case AdminDurableWorkerHealth.WORKING:
      return 'working';
    case AdminDurableWorkerHealth.UNCONFIRMED:
      return 'unconfirmed';
    case AdminDurableWorkerHealth.STALLED:
      return 'stalled';
    default:
      return 'unavailable';
  }
}

export async function getAdminSystemInfo(
  config: AdminDiagnosticsAPIConfig,
  options: { signal?: AbortSignal } = {}
): Promise<AdminSystemInfo> {
  const { client, headers } = adminDiagnosticsClient(config);
  const response = await client.getSystemInfo(
    {},
    { headers, ...(options.signal ? { signal: options.signal } : {}) }
  );
  const systemInfo = response.systemInfo;
  const cleanup = response.assetCleanup;
  const cleanupAvailable =
    cleanup != null &&
    cleanup.health !== AdminAssetCleanupHealth.UNSPECIFIED &&
    cleanup.health !== AdminAssetCleanupHealth.UNAVAILABLE;

  return {
    connection: {
      connected: systemInfo?.connection?.connected ?? false,
      serverId: systemInfo?.connection?.serverId ?? '',
      serverName: systemInfo?.connection?.serverName ?? '',
      version: systemInfo?.connection?.version ?? '',
      maxPayload: Number(systemInfo?.connection?.maxPayload ?? 0),
      rtt: systemInfo?.connection?.rtt ?? ''
    },
    account: {
      memory: Number(systemInfo?.account?.memory ?? 0),
      memoryUsed: Number(systemInfo?.account?.memoryUsed ?? 0),
      storage: Number(systemInfo?.account?.storage ?? 0),
      storageUsed: Number(systemInfo?.account?.storageUsed ?? 0),
      streams: systemInfo?.account?.streams ?? 0,
      streamsUsed: systemInfo?.account?.streamsUsed ?? 0,
      consumers: systemInfo?.account?.consumers ?? 0,
      consumersUsed: systemInfo?.account?.consumersUsed ?? 0
    },
    accountAvailable: systemInfo?.account != null,
    natsAvailable: systemInfo?.nats != null,
    nats: {
      totalMessages: Number(systemInfo?.nats?.totalMessages ?? 0),
      totalBytes: Number(systemInfo?.nats?.totalBytes ?? 0),
      totalConsumerPending: Number(systemInfo?.nats?.totalConsumerPending ?? 0),
      totalAckPending: systemInfo?.nats?.totalAckPending ?? 0,
      streams: (systemInfo?.nats?.streams ?? []).map((stream) => ({
        name: stream.name,
        description: stream.description,
        subjects: [...stream.subjects],
        storage: stream.storage,
        messages: Number(stream.messages),
        bytes: Number(stream.bytes),
        firstSequence: stream.firstSequence,
        lastSequence: stream.lastSequence,
        consumerCount: stream.consumerCount,
        replicas: stream.replicas,
        clusterLeader: stream.clusterLeader
      })),
      consumers: (systemInfo?.nats?.consumers ?? []).map((consumer) => ({
        stream: consumer.stream,
        name: consumer.name,
        durable: consumer.durable,
        filterSubject: consumer.filterSubject,
        filterSubjects: [...consumer.filterSubjects],
        ackPolicy: consumer.ackPolicy,
        pullBased: consumer.pullBased,
        pushBound: consumer.pushBound,
        pending: Number(consumer.pending),
        ackPending: consumer.ackPending,
        redelivered: consumer.redelivered,
        waiting: consumer.waiting,
        deliveredConsumerSequence: consumer.deliveredConsumerSequence,
        deliveredStreamSequence: consumer.deliveredStreamSequence,
        ackFloorConsumerSequence: consumer.ackFloorConsumerSequence,
        ackFloorStreamSequence: consumer.ackFloorStreamSequence
      }))
    },
    stats: {
      userCount: systemInfo?.stats?.userCount ?? 0,
      channelRoomCount: systemInfo?.stats?.channelRoomCount ?? 0,
      dmRoomCount: systemInfo?.stats?.dmRoomCount ?? 0
    },
    statsAvailable: systemInfo?.stats != null,
    assetCleanup: {
      available: cleanupAvailable,
      health: assetCleanupHealth(cleanup?.health),
      pendingCount: Number(cleanup?.pendingCount ?? 0),
      oldestPendingAt: cleanup?.oldestPendingAt?.toDate() ?? null,
      passInProgress: cleanup?.passInProgress ?? false,
      lastPassAt: cleanup?.lastPassAt?.toDate() ?? null,
      lastSuccessfulPassAt: cleanup?.lastSuccessfulPassAt?.toDate() ?? null,
      updatedAt: cleanup?.updatedAt?.toDate() ?? null,
      lastPassFailed: cleanup?.lastPassFailed ?? false,
      lastInspectedSequence: cleanup?.lastInspectedSequence ?? '0',
      latestDeletionSequence: cleanup?.latestDeletionSequence ?? '0'
    },
    durableWorkers: (response.durableWorkers ?? []).map((worker) => ({
      key: worker.key,
      health: durableWorkerHealth(worker.health),
      pendingCount: worker.pendingCount,
      ackPendingCount: worker.ackPendingCount,
      waitingCount: worker.waitingCount,
      redeliveredCount: worker.redeliveredCount,
      lastDeliveredSequence: worker.lastDeliveredSequence,
      ackFloorSequence: worker.ackFloorSequence
    })),
    projections: response.projections.map((projection) => ({
      key: projection.key,
      name: projection.name,
      subjects: [...projection.subjects],
      started: projection.started,
      startupDurationSeconds: projection.startupDurationSeconds ?? null,
      lastAppliedSequence: projection.lastAppliedSequence,
      matchingStreamSequence: projection.matchingStreamSequence,
      streamLastSequence: projection.streamLastSequence,
      lag: Number(projection.lag),
      failed: projection.failed,
      failedSequence: projection.failedSequence,
      failure: projection.failure,
      entryCount: Number(projection.entryCount),
      estimatedBytes: Number(projection.estimatedBytes),
      averageEntryBytes: Number(projection.averageEntryBytes),
      metrics: projection.metrics.map((metric) => ({
        name: metric.name,
        value: Number(metric.value),
        bytes: Number(metric.bytes)
      }))
    })),
    projectionsAvailable: response.projectionsAvailable ?? true
  };
}
